arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java

```diff
@@ -0,0 +1,325 @@
+     *                          a maximum of {@code batchSize} rows.
+     * @param reuseContainers   If set to {@code false}, every {@link Iterator#next()} call creates
+     *                          new instances of Arrow vectors.
+     *                          If set to {@code true}, the Arrow vectors in the previous
+     *                          {@link Iterator#next()} may be reused for the data returned
+     *                          in the current {@link Iterator#next()}.
+     *                          This option avoids allocating memory again and again.
+     *                          Irrespective of the value of {@code reuseContainers}, the Arrow vectors
+     *                          in the previous {@link Iterator#next()} call are closed before creating
+     *                          new instances if the current {@link Iterator#next()}.
+     */
+    VectorizedCombinedScanIterator(
+        CloseableIterable<CombinedScanTask> tasks,
+        Schema expectedSchema,
+        String nameMapping,
+        FileIO io,
+        EncryptionManager encryptionManager,
+        boolean caseSensitive,
+        int batchSize,
+        boolean reuseContainers) {
+      this.fileTasks = StreamSupport.stream(tasks.spliterator(), false)
+          .map(CombinedScanTask::files)
+          .flatMap(Collection::stream)
+          .collect(Collectors.toList());
+      this.fileItr = fileTasks.iterator();
+      this.inputFiles = Collections.unmodifiableMap(fileTasks.stream()
+          .flatMap(fileScanTask -> Stream.concat(Stream.of(fileScanTask.file()), fileScanTask.deletes().stream()))
+          .map(file -> EncryptedFiles.encryptedInput(io.newInputFile(file.path().toString()), file.keyMetadata()))
+          .map(encryptionManager::decrypt)
+          .collect(Collectors.toMap(InputFile::location, Function.identity())));
+      this.currentIterator = CloseableIterator.empty();
+      this.expectedSchema = expectedSchema;
+      this.nameMapping = nameMapping;
+      this.caseSensitive = caseSensitive;
+      this.batchSize = batchSize;
+      this.reuseContainers = reuseContainers;
+    }
+
+    @Override
+    public boolean hasNext() {
+      try {
+        while (true) {
+          if (currentIterator.hasNext()) {
+            this.current = currentIterator.next();
+            return true;
+          } else if (fileItr.hasNext()) {
+            this.currentIterator.close();
+            this.currentTask = fileItr.next();
+            this.currentIterator = open(currentTask);
+          } else {
+            this.currentIterator.close();
+            return false;
+          }
+        }
+      } catch (IOException | RuntimeException e) {
+        if (currentTask != null && !currentTask.isDataTask()) {
+          throw new RuntimeException(
+              "Error reading file: " + getInputFile(currentTask).location() +
+                  ". Reason: the current task is not a data task, i.e. it cannot read data rows. " +
```

Comment:
This looks incorrect to me. A data task is one that produces its own rows. For example, a static task with version history for the history table is a data task. This description seems to say the opposite is a problem, that if a task is not a data task it can't be read. That's not correct.

Reply:
Ryan, I need some clarification from you.
(For reference, this code is inspired from https://github.com/apache/iceberg/blob/master/spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java#L100.) In the spark code, the same check is done to do some error logging. If this check is not correct, we can get rid of this exception (the exception is re-raised later in the catch block). Otherwise, let me know if I should correct the exception message only.

Reply:
I think the problem is only the exception message. Let's get rid of the reason since it isn't accurate and go with the same thing that the `BaseDataReader` does. I think that this should also _not_ change the exception. Instead, I think it is a good idea to log the problem with the file location and then re-throw the original exception. Wrapping the exception makes it so any problem that occurs in the block can't be caught and handled.